### ISSUE TITLE
OnDisconnected corrupts state of ResultBox<bool>.store when PING on top of the message queue

### ISSUE THREAD

**Issue by bencican** [StackExchange/StackExchange.Redis#237]
Hi Mark,

I was chasing down a race today where seemingly put transaction Commits returned false even though these went through just fine from redis server logs.

Debugging further into it PhysicalBridge.OnDisconnect turned out to be the root cause as when a PING message is on the top of the queue it will get completed twice. I marked them in the code snippet below but basically what happens is that it gets completed the first time because it's a ping... and then second time as part of ChangeState(State.Disconnected) -> AbortUnsent().

This is consistent with the scenario I was encountering where we had a transaction on a disconnected multiplexer which failed and subsequent transaction on a connected multiplexer which succeeded but returned result as though it failed too.

```c#
    internal void OnDisconnected(ConnectionFailureType failureType, PhysicalConnection connection, out bool isCurrent, out State oldState)
    {
        Trace("OnDisconnected");

        // if the next thing in the pipe is a PING, we can tell it that we failed (this really helps spot doomed connects)
        // note that for simplicity we haven't removed it from the queue; that's OK
        int count;
        var ping = queue.PeekPing(out count);
        if (ping != null)
        {
            Trace("Marking PING as failed (queue length: " + count + ")");
            ping.Fail(failureType, null);

            // First completion <--------------------------------------------
            CompleteSyncOrAsync(ping);
        }
        oldState = default(State); // only defined when isCurrent = true
        if (isCurrent = (physical == connection))
        {
            Trace("Bridge noting disconnect from active connection" + (isDisposed ? " (disposed)" : ""));

            // Second completion <--------------------------------------------
            oldState = ChangeState(State.Disconnected);
            physical = null;

            if (!isDisposed && Interlocked.Increment(ref failConnectCount) == 1)
            {
                GetConnection(null); // try to connect immediately
            }
        }
```

Two completions cause the PING message to be completed twice which in turn completes the underlying ResultBox<bool> twice adding a duplicate to the ResultBox<bool>.store and thus corrupting the state of the store and upcoming messages based on how unlucky you get since now a single ResultBox<bool> will be potentially updated completed from more than one message callback at any given time.

Please let me know if there's any more info needed. I don't have a clean repro atm but I think the issue is easy enough to follow and understand.

**Comment by nmehlei** [comment]
Could this be possibly related to https://github.com/StackExchange/StackExchange.Redis/issues/205 ? This ticket describes a case where during high utilization scenarios (and/or timeframes with connection interruptions, like here) the KeyExists command returns an incorrect boolean value. If the returned bool value was instead originated from the Ping message than this could quite possibly explain it.

**Comment by kylestittleburg** [comment]
+1 on this issue, I think we're seeing it during out load tests.  #275 has a relatively reliable repro test, for me at least, especially if I bump the number of threads up.

Is the intent of `store` in `ResultBox<T>` to be a performance optimization to save on construction/cleanup costs for `ResultBox` objects?

**Comment by NickCraver** [comment]
Fixed in #279, closing out to cleanup!
